src/main/java/com/minecolonies/coremod/inventory/GuiRack.java

```diff
@@ -88,10 +117,41 @@ public GuiRack(final InventoryPlayer parInventoryPlayer, final TileEntityRack ti
     protected void drawGuiContainerBackgroundLayer(final float partialTicks, final int mouseX, final int mouseY)
     {
         GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
-        this.mc.getTextureManager().bindTexture(CHEST_GUI_TEXTURE);
+        this.mc.getTextureManager().bindTexture(getCorrectTextureForSlots(inventoryRows));
         final int i = (this.width - this.xSize) / 2;
         final int j = (this.height - this.ySize) / 2;
-        this.drawTexturedModalRect(i, j, 0, 0, this.xSize, this.inventoryRows * SLOT_OFFSET + SLOT_OFFSET-1);
-        this.drawTexturedModalRect(i, j + this.inventoryRows * SLOT_OFFSET + SLOT_OFFSET-1, 0, TEXTURE_OFFSET, this.xSize, TEXTURE_HEIGHT);
+
+        if (inventoryRows < SLOTS_EACH_ROW)
+        {
+            drawModalRectWithCustomSizedTexture(i, j, 0, 0, this.xSize, this.inventoryRows * SLOT_OFFSET + SLOT_OFFSET - 1, TEXTURE_SIZE, TEXTURE_SIZE);
+            drawModalRectWithCustomSizedTexture(i, j + this.inventoryRows * SLOT_OFFSET + SLOT_OFFSET - 1, 0,
+                    TEXTURE_OFFSET, this.xSize, TEXTURE_HEIGHT, TEXTURE_SIZE, TEXTURE_SIZE);
+        }
+        else
+        {
+            final int textureOffset = TEXTURE_OFFSET - EXTRA_OFFSET;
+            drawModalRectWithCustomSizedTexture(i, j, 0, 0, (this.xSize * 3) / 2, this.inventoryRows * SLOT_OFFSET + SLOT_OFFSET - 1, TEXTURE_SIZE, TEXTURE_SIZE);
+            drawModalRectWithCustomSizedTexture(i,
+                    j + Math.min(SLOTS_EACH_ROW, this.inventoryRows) * SLOT_OFFSET + SLOT_OFFSET - 1, 0,
+                    textureOffset, (this.xSize * 3) / 2, TEXTURE_HEIGHT + EXTRA_HEIGHT, TEXTURE_SIZE, TEXTURE_SIZE);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 3 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)